ibis/common/graph.py

```diff
@@ -28,6 +29,18 @@ def __children__(self, filter=None):
     def __rich_repr__(self):
         return zip(self.__argnames__, self.__args__)
 
+    def branches(self):
+        stack = [(self, [])]
+
+        while stack:
+            node, path = stack.pop()
+
+            if children := node.__children__():
+                for child in reversed(children):
+                    stack.append((child, path + [node]))
```

Comment:
This is going to visit some child patterns before all parents are visited:

```mermaid
flowchart TD
    A[1]
    A --> B[2]
    A --> C[4]
    B --> D[3]
    C --> D
```

Is that intentional?